package.json

```diff
@@ -60,6 +60,7 @@
     "graphql": "^15.3.0",
     "graphql-tag": "^2.11.0",
     "lodash-es": "^4.17.15",
+    "resize-observer-polyfill": "^1.5.1",
```

Comment:
We already have our own implementation of this which we use on all other visualizations. I can't remember the exact reason why we didn't use something like resize observer, but I do remember looking at it. Example usage from cartesian:
```html
<div #chartContainer class="fill-container" (htLayoutChange)="this.redraw()"></div>
```

Reply:
This is triggered for the entire window/dashboard. We can use it, but i strongly feel we should use the resizeObserver from inside this component. It would get triggered even if window/layout size doesn't change. Like if we have gauge and another component inside a flex container, and the other component is removed with NgIf, layout change wouldn't get triggered.

Reply:
htLayoutChange uses the window as its root, but is hierarchical - each new layout change directive creates a new scope. Even if we eventually switch to a lib (don't think it's needed personally), that should be wrapped in this service for simplified use.

Reply:
Right now we have wrapped it up into an rxjs observable easily. Do we still need a service ? If yes, what would we gain?

Reply:
So trying to page back to why we implemented it ourselves, I _think_ other than the feature not having widespread support, the issue is that the way we often have implemented layouts (particularly with d3 stuff), a resize wouldn't trigger the way we'd want. For example, we may have done a measurement to set the svg's width, so when other parts of the dom are hidden, this component hasn't actually changed size - it's already explicitly sized. That also would applied to a parent div which got its size based on its child content.

Reply:
Also, LayoutChangeService seems to be doing a lot or processing. I would prefer to not depend on an outside service and keep things local to this component

Reply:
So right now, Resize Observer is support by all major browsers. Consider, the svg's are in angular context, I don't see any harm as such/

Reply:
Other than the point above, I think the usage of using an output vs building an observable ourselves just feels a little cleaner and more declarative (and should tie into change detection, too - although most things that would trigger a resize already do).

Reply:
It looks like the browser support got added since our layout change was written, but even though more code, given the explicit trigger it should be doing much less work than a resize observer which needs to continually measure. The performance and support are immaterial though, if the resize observer doesn't actually do what we want as I described above.

Reply:
Resize Observer has in built browser support. Isn't it designed in a reactive fashion with native support? I disagree that it is continuously measuring dom in a brute force approach. 

While LayoutService may have advantages in other cases, for this particular component, AFAIK, it is missing the functionality to trigger a layout change if container dimension changes due to removing a sibling element (with no window resize). Can you confirm this please?

This also feels like instead of calling `addEventListener` or `fromEvent`, we are watching for an event through an external service.

Reply:
> Resize Observer has in built browser support. Isn't it designed in a reactive fashion with native support? I disagree that it is continuously measuring dom in a brute force approach.

Fair, the native implementation should be performant, the polyfill likely not.

> While LayoutService may have advantages in other cases, for this particular component, AFAIK, it is missing the functionality to trigger a layout change if container dimension changes due to removing a sibling element (with no window resize). Can you confirm this please?

That generally wouldn't be triggered by this component (since it never resizes itself), but by a parent that would remove a sibling element (for example, the left nav or our panel trigger this). Since our visualizations are generally used as widgets, we don't have very many places that need to trigger. The gist of it is that layout change model abstracts the source - we can always use resize observer within layout change if we later find the need, and combine it with any other inputs too.

As a related aside - because this component is using a relative size (100% - rather than an explicit size), a resize observer works, but doesn't that have other implications? For example, because this is an SVG, doesn't that imply the text will scale based on the gauge size? While a pain, this is the main reason most of our visualizations use explicit sizes, so we can pick and choose what scales.

Reply:
I would prefer to use resize Observer. https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/948 will also be in soon. I could wait for it too. 

I tried to use LayoutChangeService and replaced it in the design I have above. The first time It tries to render (Default, without layout change), the root element gets a wrong width. This was not happening with the resize observer we used before.

Reply:
Is this the problem you were mentioning above?

Reply:
> I tried to use LayoutChangeService and replaced it in the design I have above. The first time It tries to render (Default, without layout change), the root element gets a wrong width. This was not happening with the resize observer we used before.

Unclear what you mean - how would adding a (non-structural) directive change the width? Can you post a separate PR where you try the layout change directive? This is the method we use on every other svg, so if there are issues, we should raise them.

> Is this the problem you were mentioning above?

The problem I'm referencing above is that a responsive svg scales evenly for all elements, text included. That's not behavior we want with our visualizations.

Reply:
https://github.com/hypertrace/hypertrace-ui/pull/339 here you go . The problem is the dom first has a larger width. So it rendering origin is off. Then during the next trigger, it gets the correct width. Wondering why it is happening. With Resize observable, this doesn't happen.

Reply:
My guess is that the same issue that appears in both - you're building the measuring observable in the constructor, which is too early to measure as the dom hasn't been laid out yet. In this form, you're debouncing the resize observable which means the first draw doesn't actually happen at that moment, and is likely delayed until after drawing. Try assigning it in afterViewInit instead.

Reply:
Right. ViewInit was somehow adding a lot of delay I have tried that already. I am using debounce to fix this.

Reply:
Updated this PR. Disregard the other one